docfx/articles/memory.md

```diff
@@ -40,14 +35,67 @@ using `using` in C# or explicit calls to `System.IDisposable.Dispose()`.
 
 👎 Temporaries are not covered by this approach, so to maximize the benefit, you may have to store all temporaries to variables and dispose.
 
-> NOTE: Disposing a tensor only releases the underlying storage if this is the last
-> live TorchTensor which has a view on that tensor -- the native runtime does reference counting of tensors.
+__Note__: Even with this approach, it is a good idea to place a call to `GC.Collect()` after each mini-batch of data. There may be temporaries that were overlooked, or inconvenient to pull out, or ones where the lifetime was unclear; calling `GC.Collect()` will catch them.
+
+
+### Returning Fresh References
+
+It is important to understand that all TorchSharp tensors are simple wrappers around a C++ tensor. The 'Handle' property on `Tensor` is a pointer to a C++ `shared_ptr<at::Tensor>`. When a Tensor is created and returned to .NET, the reference count is incremented. When you call `Dispose()` on the tensor, it is decremented.
+
+To enable this technique, all Tensor operations should return a fresh reference to a Tensor, if not a fresh Tensor. This is true even for in-place, destructive operations like `add_()`, which overwrites the underlying native tensor with data, but still returns a Tensor with a handle that is different from the one going in.
```

Comment:
Suggestion:

```
It is important to understand that all TorchSharp "tensors" (type Tensor) are actually "tensor references", referring to a C++ tensor. When a C++ tensor is created and returned to .NET as a tensor reference, the reference count on the C++ tensor is incremented. When you call `Dispose()` on the TorchSharp tensor reference (that is, type Tensor), it is decremented. If the tensor reference is disposed, the decrement happens implicitly.

To enable this technique, all operations that return one or more TorchSharp `Tensor`s should return "fresh" Tensor objects (though that doesn't necessarily mean freshly copied C++ Tensors). This is true even for in-place, destructive operations like `add_()`, which overwrites the underlying native tensor with data, but still returns a fresh Tensor reference to that same tensor.  

Thus, when you write methods and functions that take and produce Tensor, for example in the `forward()` method of a model, you should always make sure to return a fresh reference. Most of the time, this happens automatically, because the last action of your code will normally be to call another tensor function, but there are cases when it's not, especially when returning input tensors or tensors stored in some lookaside table.
```

NOTE: should the getters of properties of models return fresh tensor handles?  I suppose not. But then we should explicitly call out when this is not done, e.g. `model.weights` and `model.parameters()` etc.

Reply:
We could perhaps systematically call them "Tensor aliases".  The use of "Tensor reference" is really confusing given the presence of "reference" and "by-ref(erence)" in .NET languages

Reply:
Sorry if I bother you. For in-place functions, I guess we can dispose the original handle and assign the new handle to the C# Tensor, so that this behavior will be transparent to TorchSharp users and "x.add_().mul_().add_()" is still good practice.